### DISCUSSION TITLE
Best practice: How to prevent leeching?

### DISCUSSION THREAD

**Discussion by tobwen** [protomaps/PMTiles#144] [Q&A] [answered]
## problem
In my eyes, PMTiles have a a small disadvantage: you can leeche the complete file archive and thus get all the data in one swoop.

## alternatives
With MVT (or PNG/WMS/WMTS in terms of raster), this is much more difficult because thousands of queries have to be sent to the server. This can be filtered out reasonably well by rate limiting.

## assumption
Let's assume that the PMTiles contain data that is properitary and you want to prevent leeching. 

## question
How do you go about it? Limit the size of the range? In NGINX you would probably have to tinker something in LUA here, right?

## hint
The argument that PMTiles are not suitable for properitary data is understandable, but should not stand in the way of this discussion.

**Comment by thomasknowles** [comment, marked as answer]
Some solutions for AWS and cloudflare can be found here (by the author):

https://protomaps.com/docs/cdn

It requires extra work to implement.

**Reply by tobwen** [reply]
Sure, that's a workaround. But then add server/service where the data resides (customers often don't like 3rd parties). For updates, the entire cache must flushed - unless you selectively delete the changed tiles from the cache (but many CDN providers do not allow this).

**Reply by bdon** [reply]
Hi @tobwen - the solution that @thomasknowles linked is the intended solution to this. It does add some complexity, but the end result of CDN-cached tiles is generally the desired one for high traffic public deployments.

I understand what you mean related to caching, however. There isn't a good solution right now for the combination of unleechable tiles + no caching layer. my recommendation is to set a CDN time-to-live - e.g. if your tileset is updated once a day, set a TTL of 1 hour. This is controlled via environment variable for Cloudflare or Cloudfront settings on AWS. If your application requires very frequent updates a different software stack may be more appropriate like PostgreSQL instead of static tilesets.
